### PR TITLE
Api docs

### LINKED ISSUE #149: Generate API documentation
The code has lots of docstrings it would be nice to see them renderend somewhere like the mkdir api docs.

### PR DESCRIPTION
**Pull request by Peter9192**
- Generates API docs using mkdocstrings and mkdocs gen-pages, with slightly tweaked css
- Cleans up docstrings and makes them consistent accross datasets
- Auto-generate documentation of cli recipes similar to api generation
- Moves dataset recipes from tests to src/recipes/datasets

To do (in follow-up):
- Transpose documentation? Instead of (notebook/api/cli -> dataset) list like (dataset -> notebook/api/cli)
- Split up example recipes into individual datasets?

closes #149

### CODE REVIEW

**Comment by Peter9192** [comment]
https://github.com/pytest-dev/pytest/issues/4030

Running doctest as standalone behaves differently then running it through pytest:

```bash
# With doctest
python -m doctest -v src/springtime/datasets/NPNPhenor.py

# With pytest
pytest --doctest-modules src/springtime/datasets/NPNPhenor.py -v
```

Pytests correctly limits terminal width to 80 (the default value), whereas doctest let's it depend on terminal width. Thus, to get reproducible behaviour, the best option is to always run doctests through pytest. We have two options for the pandas repr behaviour:

* Option 1: Use pytest with default of 80 chars
* Option 2: Use pytest fixture to customize the printed dfs:

```python
import pandas as pd
import pytest

@pytest.fixture(autouse=True, scope='session')
def pandas_terminal_width():
    pd.set_option('display.width', None)
    pd.set_option("display.max_columns", 7)
    pd.set_option("display.max_colwidth", 20)
```

The fixture needs to be in `conftest.py` in the same dir as the module that's being tested.

**Comment by Peter9192** [comment]
To get copy-pastable output from doctest: `--doctest-report none`
https://docs.pytest.org/en/7.1.x/how-to/doctest.html#output-format

**Comment by Peter9192** [comment]
> * Option 1: Use pytest with default of 80 chars
> 
> * Option 2: Use pytest fixture to customize the printed dfs:

Prefer option 1. More reproducible, and also keeps linter happy :-)

**Comment by Peter9192** [comment]
Code blocks formatted like this render nicely and also work with doctest:

```
class A:
    """
    Test

    Example:

        ```pycon
        >>> import time
        >>>
        >>> print('something')
        something

        ```

    """
```

Notes: 
- "Example" and "Examples" behave differently
- Whiteline at the end is necessary to break out of the output block
- "pycon" is syntax highlight for python console (https://pygments.org/docs/lexers/#pygments.lexers.python.PythonConsoleLexer)
- Modifying copy behaviour is possible but doesn't feel nice (https://github.com/squidfunk/mkdocs-material/issues/3647)

**Comment by Peter9192** [comment]
Having a hard time deciding whether to keep the example notebooks, doctests (or api docs in general), or both. There's signification overlap, but they also have their unique advantages.

- Notebooks are easier to work with locally to fix them if they're broken
- Notebooks have a nicer rendering of input vs output (doctest has these `>>>`, notebook have separate cells)
- Doctests serve both as examples with output illustration, and tests
- Notebooks could also be tested, e.g. with [nbval](https://nbval.readthedocs.io/en/latest/), [nbmake](https://pypi.org/project/nbmake/), or [pytest-jupyter](https://pypi.org/project/pytest-jupyter/)
- Doctests are closer to the source code
- In both cases, execution time is a challenge (and dependency on server calls)

**Comment by Peter9192** [comment]
Decided to keep both apidocs examples and notebooks.

File: docs/notebooks/datasets_modis_appeears.ipynb
Comment (sverhoeven) [inline docs/notebooks/datasets_modis_appeears.ipynb:1]: The notebook is not rendered in mkdocs

Comment (sverhoeven) [reply docs/notebooks/datasets_modis_appeears.ipynb:1]: Fixed in 6271fa7cfff819901761c94e6bfbaa73e32e7c19

**Review by sverhoeven — APPROVED** [review]
I have spunoff https://github.com/phenology/springtime/issues/182 to continue doc improvements.

I fixed most of my own problems/suggestions. I hope you like my additions.
I learned a lot about mkdocs during my review.
Having this extra nesting in src/datasets might bite us in the ass when we have a dataset that is both meteo and satellite, but will solve that if that occurs.

Nice work figuring out how to render everything.

File: docs/gen_cliref_datasets.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+    with mkdocs_gen_files.open(full_doc_path, "w") as fd:
+        fd.write("```yaml\n")
+        fd.writelines(recipe)
+        fd.write("\n```\n")
+
+    mkdocs_gen_files.set_edit_path(full_doc_path, path)
+
+with mkdocs_gen_files.open("clireference/SUMMARY.md", "w") as nav_file:
+    nav_file.writelines(nav.build_literate_nav())
```
[the remark this answers is not in the document]
Comment (sverhoeven) [reply docs/gen_cliref_datasets.py:31]: Could try static file instead of gen_cliref_datasets.py

**Comment by Peter9192** [comment]
Thanks for the review and adding your own suggestions! We now seem to have a new issue with displaying type inheritance:

image

I think the issue appeared only recently due to the latest release of griffe including https://github.com/mkdocstrings/griffe/pull/170. 

ps in the screenshot also the internal links appear to have broken.

**Comment by Peter9192** [comment]
Okay, after a lot of tinkering I figured out that the order of the docstrings is what causes this weird behaviour.

Option 1: first own attributes, then inherited attributes --> repeated type of last own attribute in all inherited docstrings
Option 2: first inherited attributes, then own attributes --> empty type for inherited docstrings

File: src/springtime/datasets/abstract.py [hunk 2]
```diff
@@ -9,13 +17,19 @@
...
 
 class Dataset(BaseModel, ABC):
+    """Base class for springtime datasets.
+
+    Attributes:
+        dataset: The name of the dataset.
+        years: timerange. For example years=[2000, 2002] downloads data for
+            three years.
+        resample: Resample the dataset to a different time resolution. If None,
+            no resampling.
```
[the remark this answers is not in the document]
Comment (Peter9192) [reply src/springtime/datasets/abstract.py:27]: It had to do with the order of the fields, apparently when you docstring inherited fields first then you get your behaviour, if you put them last, you get mine. Final optimal solution could come from https://github.com/mkdocstrings/python/issues/58

File: docs/gen_cliref_datasets.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Peter9192) [reply docs/gen_cliref_datasets.py:52]: Done in https://github.com/phenology/springtime/pull/164/commits/868fe1b1ca49f8a5d064f09ad53623c710411fb5

File: src/springtime/recipes/simple_usecase.yaml
[the remark this answers is not in the document]
Comment (Peter9192) [reply src/springtime/recipes/simple_usecase.yaml:1]: let's discuss further in #181

**Comment by sverhoeven** [comment]
Ok so you went with option 2, so we are not displaying incorrect data just empty types.

**Comment by Peter9192** [comment]
Indeed.

**Pull request merged**